File: codeo/0x53/main.cpp
#include <iostream>
#include <vector>
#include <map>
#include <iterator>

using namespace std;

struct NODE
{
    vector<int> edges;
    bool visited;
    bool explore;
};

map<int, NODE> graph;

bool dfs(NODE node) {
    if (node.visited) return false;
    if (node.explore) return true;
    node.explore = true;
    for (int i = 0; i < node.edges.size(); i++) {
        if (!graph[node.edges[i]].visited && dfs(graph[node.edges[i]])) return true;
    }
    node.explore = false;
    node.visited = true;
    return false;
}

bool hasCycles() {
    map<int, NODE>::iterator it = graph.begin();
    while (it != graph.end()) {
        NODE node = it->second;
        if (!node.visited && dfs(node)) return true;
        it++;
    }
    return false;
}
    

int main() {
    
    int n, m;
    cin >> n >> m;
    int u, v;
    for (int i = 0; i < m; i++) {
        cin >> u >> v;
        graph[u].edges.push_back(v);
    }
    if (hasCycles()) {
        cout << "Yes" << endl;
    } else {
        cout << "No" << endl;
    }
    return 0;
}

File: codeo/0x53/main.py
graph = {}

# Depth First Search
def dfs(node):
    if node["visited"] is True:
        return False
    if node["explore"] is True:
        return True
    node["explore"] = True
    for edge in node["edges"]:
        edge = graph[edge]
        if edge["visited"] is False and dfs(edge) is True:
            return True
    node["explore"] = False
    node["visited"] = True
    return False

def has_cycles():
    for node in graph.values():
        if node["visited"] is False and dfs(node) is True:
            return True
    return False

n, m = [int(x) for x in input().split()]

for i in range(n):
    graph[i] = {
        "edges": set(),
        "visited": False,
        "explore": False
    }

for i in range(m):
    u, v = [int(x) for x in input().split()]
    graph[u]["edges"].add(v)

print("Yes" if has_cycles() else "No")
